lib/Transforms/TupleToStruct.cpp

```diff
@@ -0,0 +1,123 @@
+// We explicitly do not use push / pop for diagnostic in
+// order to propagate pragma further on
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/Pass.h"
+
+#include "p4mlir/Transforms/Passes.h"
+#include "p4mlir/Conversion/ConversionPatterns.h"
+#include "p4mlir/Dialect/P4HIR/P4HIR_Ops.h"
+
+
+#define DEBUG_TYPE "p4hir-tuple-to-struct"
+
+using namespace mlir;
+
+namespace P4::P4MLIR {
+#define GEN_PASS_DEF_TUPLETOSTRUCT
+#include "p4mlir/Transforms/Passes.cpp.inc"
+}
+
+using namespace P4::P4MLIR;
+
+namespace {
+
+struct TupleToStructPass
+    : public P4::P4MLIR::impl::TupleToStructBase<TupleToStructPass> {
+    TupleToStructPass() = default;
+    void runOnOperation() override;
+};
+
+
+class TupleToStructTypeConverter: public P4HIRTypeConverter {
+ public:
+    explicit TupleToStructTypeConverter() {
+
+        addConversion([](TupleType tupleType)-> Type {
+            MLIRContext *ctx = tupleType.getContext();
+            ArrayRef<Type> elementTypes = tupleType.getTypes();
+
+            SmallVector<P4::P4MLIR::P4HIR::FieldInfo> structFiedls;
+            structFiedls.reserve(elementTypes.size());
+
+            for(unsigned i=0; i< elementTypes.size(); i++){
+                mlir::StringAttr name = mlir::StringAttr::get(ctx, "elemet_" + std::to_string(i));
+                P4::P4MLIR::P4HIR::FieldInfo field = P4::P4MLIR::P4HIR::FieldInfo(name, elementTypes[i]);
+                structFiedls.push_back(field);
+            }
+            llvm::StringRef name("_tupletoStruct");
+            return P4::P4MLIR::P4HIR::StructType::get(ctx, name, structFiedls);
+        });
+    }
+};
+
+struct TupleOpConversion : public OpConversionPattern<P4::P4MLIR::P4HIR::TupleOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(P4::P4MLIR::P4HIR::TupleOp op, OpAdaptor adaptor,
+                                ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    Type converted = getTypeConverter()->convertType(op.getType());
+    auto structType = dyn_cast<P4::P4MLIR::P4HIR::StructType>(converted);
+
+    if(!structType)
+      return rewriter.notifyMatchFailure(op, "Type conversion failed to struct");
+
+    Value newStruct = rewriter.create<P4::P4MLIR::P4HIR::StructOp>(loc, structType, adaptor.getOperands());
+    rewriter.replaceOp(op, newStruct);
+    return success();
+
+   }
+};
+
+struct TupleExtractOpConversion : public OpConversionPattern<P4::P4MLIR::P4HIR::TupleExtractOp> {
```

Comment:
Why these repeating `P4::P4MLIR`? You're already `using` this namespace above. Same in other places